File: app/src/main/java/de/westnordost/streetcomplete/quests/localized_name/AddRoadNameForm.java
package de.westnordost.streetcomplete.quests.localized_name;

import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import javax.inject.Inject;

import de.westnordost.osmapi.map.data.LatLon;
import de.westnordost.streetcomplete.Injector;
import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.data.meta.Abbreviations;
import de.westnordost.streetcomplete.data.meta.AbbreviationsByLocale;
import de.westnordost.streetcomplete.data.osm.ElementGeometry;
import de.westnordost.streetcomplete.quests.localized_name.data.RoadNameSuggestionsDao;


public class AddRoadNameForm extends AddLocalizedNameForm
{

	public static final String
			NO_PROPER_ROAD = "no_proper_road",
			WAY_ID = "way_id",
			WAY_GEOMETRY = "way_geometry";

	public static final int IS_SERVICE = 1, IS_LINK = 2, IS_TRACK = 3;

	@Inject AbbreviationsByLocale abbreviationsByLocale;
	@Inject RoadNameSuggestionsDao roadNameSuggestionsDao;

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
							 Bundle savedInstanceState)
	{
		View view = super.onCreateView(inflater, container, savedInstanceState);

		Injector.instance.getApplicationComponent().inject(this);

		View contentView = setContentView(R.layout.quest_localizedname);

		addOtherAnswers();

		initLocalizedNameAdapter(contentView, savedInstanceState);

		return view;
	}

	private void addOtherAnswers() {
		addOtherAnswer(R.string.quest_name_answer_noName, this::selectNoStreetNameReason);
		addOtherAnswer(R.string.quest_streetName_answer_cantType, this::showKeyboardInfo);
	}

	@Override
	protected AddLocalizedNameAdapter setupNameAdapter(ArrayList<LocalizedName> data, Button addLanguageButton) {
		return new AddLocalizedNameAdapter(
			data, getActivity(), getPossibleStreetsignLanguages(),
			abbreviationsByLocale, getRoadnameSuggestions(), addLanguageButton);
	}

	private List<Map<String, String>> getRoadnameSuggestions()
	{
		ElementGeometry geometry = getElementGeometry();
		if(geometry == null || geometry.polylines == null || geometry.polylines.isEmpty())
		{
			return new ArrayList<>();
		}
		List<LatLon> points = geometry.polylines.get(0);
		List<LatLon> onlyFirstAndLast = Arrays.asList(points.get(0), points.get(points.size()-1));

		return roadNameSuggestionsDao.getNames(onlyFirstAndLast, AddRoadName.MAX_DIST_FOR_ROAD_NAME_SUGGESTION);
	}

	@Override protected void onClickOk()
	{
		LinkedList<String> possibleAbbreviations = new LinkedList<>();
		for (LocalizedName localizedName : adapter.getData())
		{
			String name = localizedName.name;
			if(name.trim().isEmpty())
			{
				Toast.makeText(getActivity(), R.string.quest_generic_error_a_field_empty,
					Toast.LENGTH_LONG).show();
				return;
			}

			Abbreviations abbr = abbreviationsByLocale.get(new Locale(localizedName.languageCode));
			boolean containsAbbreviations = abbr != null && abbr.containsAbbreviations(name);

			if (name.contains(".") || containsAbbreviations)
			{
				possibleAbbreviations.add(name);
			}
		}

		confirmPossibleAbbreviationsIfAny(possibleAbbreviations, this::applyNameAnswer);
	}

	@Override
	protected void applyNameAnswer()
	{
		Bundle bundle = createAnswer();
		bundle.putLong(WAY_ID, getOsmElement().getId());
		bundle.putSerializable(WAY_GEOMETRY, getElementGeometry());
		applyAnswer(bundle);
	}

	private void selectNoStreetNameReason()
	{
		final String
				linkRoad = getResources().getString(R.string.quest_streetName_answer_noProperStreet_link),
				serviceRoad = getResources().getString(R.string.quest_streetName_answer_noProperStreet_service2),
				trackRoad = getResources().getString(R.string.quest_streetName_answer_noProperStreet_track2),
				noName = getResources().getString(R.string.quest_streetName_answer_noName_noname),
				leaveNote = getResources().getString(R.string.quest_streetName_answer_noProperStreet_leaveNote);

		String highwayValue = getOsmElement().getTags().get("highway");
		boolean mayBeLink = highwayValue.matches("primary|secondary|tertiary");

		final List<String> answers = new ArrayList<>(5);
		if(mayBeLink) answers.add(linkRoad);
		answers.add(serviceRoad);
		answers.add(trackRoad);
		answers.add(leaveNote);
		answers.add(noName);

		DialogInterface.OnClickListener onSelect = new DialogInterface.OnClickListener()
		{
			Integer selection = null;

			@Override public void onClick(DialogInterface dialog, int which)
			{
				if (which >= 0)
				{
					selection = which;
					((AlertDialog)dialog).getButton(DialogInterface.BUTTON_POSITIVE).setEnabled(true);
				}
				else if (which == DialogInterface.BUTTON_POSITIVE)
				{
					if(selection == null || selection < 0 || selection >= answers.size()) return;
					onAnswer();
				}
			}

			private void onAnswer()
			{
				String answer = answers.get(selection);
				if(answer.equals(leaveNote))
				{
					onClickCantSay();
				}
				else if(answer.equals(noName))
				{
					confirmNoStreetName();
				}
				else
				{
					Bundle data = new Bundle();
					int type = 0;
					if(answer.equals(linkRoad))		type = IS_LINK;
					if(answer.equals(serviceRoad))	type = IS_SERVICE;
					if(answer.equals(trackRoad))    type = IS_TRACK;
					data.putInt(NO_PROPER_ROAD, type);
					applyAnswer(data);
				}
			}
		};

		AlertDialog dlg = new AlertDialog.Builder(getActivity())
				.setSingleChoiceItems(answers.toArray(new String[0]), -1, onSelect)
				.setTitle(R.string.quest_streetName_answer_noName_question)
				.setPositiveButton(android.R.string.ok, onSelect)
				.setNegativeButton(android.R.string.cancel, null)
				.show();

		dlg.getButton(DialogInterface.BUTTON_POSITIVE).setEnabled(false);
	}

	private void confirmNoStreetName()
	{
		new AlertDialog.Builder(getActivity())
				.setTitle(R.string.quest_name_answer_noName_confirmation_title)
				.setMessage(R.string.quest_streetName_answer_noName_confirmation_description)
				.setPositiveButton(R.string.quest_name_noName_confirmation_positive, (dialog, which) ->
				{
					Bundle data = new Bundle();
					data.putBoolean(NO_NAME, true);
					applyAnswer(data);
				})
				.setNegativeButton(R.string.quest_generic_confirmation_no, null)
				.show();
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/localized_name/AddRoadNameForm.kt
package de.westnordost.streetcomplete.quests.localized_name

import android.content.DialogInterface
import android.view.View
import androidx.appcompat.app.AlertDialog

import java.util.LinkedList
import java.util.Locale

import javax.inject.Inject

import de.westnordost.streetcomplete.Injector
import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.data.meta.AbbreviationsByLocale
import de.westnordost.streetcomplete.data.osm.ElementPolylinesGeometry
import de.westnordost.streetcomplete.quests.OtherAnswer
import de.westnordost.streetcomplete.quests.localized_name.data.RoadNameSuggestionsDao
import java.lang.IllegalStateException


class AddRoadNameForm : AAddLocalizedNameForm<RoadNameAnswer>() {

    override val otherAnswers = listOf(
        OtherAnswer(R.string.quest_name_answer_noName) { selectNoStreetNameReason() },
        OtherAnswer(R.string.quest_streetName_answer_cantType) { showKeyboardInfo() }
    )

    @Inject internal lateinit var abbreviationsByLocale: AbbreviationsByLocale
    @Inject internal lateinit var roadNameSuggestionsDao: RoadNameSuggestionsDao

    init {
        Injector.instance.applicationComponent.inject(this)
    }

    override fun setupNameAdapter(data: List<LocalizedName>, addLanguageButton: View): AddLocalizedNameAdapter {
        return AddLocalizedNameAdapter(
            data, activity!!, getPossibleStreetsignLanguages(),
            abbreviationsByLocale, getRoadNameSuggestions(), addLanguageButton
        )
    }

    private fun getRoadNameSuggestions(): List<MutableMap<String, String>> {
        val polyline = (elementGeometry as ElementPolylinesGeometry).polylines.first()
        return roadNameSuggestionsDao.getNames(
            listOf(polyline.first(), polyline.last()),
            AddRoadName.MAX_DIST_FOR_ROAD_NAME_SUGGESTION
        )
    }

    override fun onClickOk(names: List<LocalizedName>) {

        val possibleAbbreviations = LinkedList<String>()
        for ((languageCode, name) in adapter.localizedNames) {
            val locale = if(languageCode.isEmpty()) countryInfo.locale else Locale(languageCode)
            val abbr = abbreviationsByLocale.get(locale)
            val containsAbbreviations = abbr?.containsAbbreviations(name) == true

            if (name.contains(".") || containsAbbreviations) {
                possibleAbbreviations.add(name)
            }
        }

        confirmPossibleAbbreviationsIfAny(possibleAbbreviations) {
            applyAnswer(RoadName(names, osmElement!!.id, elementGeometry as ElementPolylinesGeometry))
        }
    }

    private fun selectNoStreetNameReason() {
        val linkRoad = resources.getString(R.string.quest_streetName_answer_noProperStreet_link)
        val serviceRoad = resources.getString(R.string.quest_streetName_answer_noProperStreet_service2)
        val trackRoad = resources.getString(R.string.quest_streetName_answer_noProperStreet_track2)
        val noName = resources.getString(R.string.quest_streetName_answer_noName_noname)
        val leaveNote = resources.getString(R.string.quest_streetName_answer_noProperStreet_leaveNote)

        val highwayValue = osmElement!!.tags["highway"]
        val mayBeLink = highwayValue?.matches("primary|secondary|tertiary".toRegex()) == true

        val answers = mutableListOf<String>()
        if (mayBeLink) answers.add(linkRoad)
        answers.add(serviceRoad)
        answers.add(trackRoad)
        answers.add(leaveNote)
        answers.add(noName)

        val onSelect = object : DialogInterface.OnClickListener {
            var selection: Int? = null

            override fun onClick(dialog: DialogInterface, which: Int) {
                if (which >= 0) {
                    selection = which
                    (dialog as AlertDialog).getButton(DialogInterface.BUTTON_POSITIVE).isEnabled = true
                } else if (which == DialogInterface.BUTTON_POSITIVE) {
                    selection?.let {
                        if(it >= 0 && it < answers.size) onAnswer(it)
                    }
                }
            }

            private fun onAnswer(selection: Int) {
                val answer = answers[selection]
                when (answer) {
                    leaveNote -> composeNote()
                    noName    -> confirmNoStreetName()
                    else      -> {
                        applyAnswer(when(answer) {
                            linkRoad    -> RoadIsLinkRoad
                            serviceRoad -> RoadIsServiceRoad
                            trackRoad   -> RoadIsTrack
                            else        -> throw IllegalStateException()
                        })
                    }
                }
            }
        }

        val dlg = AlertDialog.Builder(activity!!)
            .setSingleChoiceItems(answers.toTypedArray(), -1, onSelect)
            .setTitle(R.string.quest_streetName_answer_noName_question)
            .setPositiveButton(android.R.string.ok, onSelect)
            .setNegativeButton(android.R.string.cancel, null)
            .show()

        dlg.getButton(DialogInterface.BUTTON_POSITIVE).isEnabled = false
    }

    private fun confirmNoStreetName() {
        AlertDialog.Builder(activity!!)
            .setTitle(R.string.quest_name_answer_noName_confirmation_title)
            .setMessage(R.string.quest_streetName_answer_noName_confirmation_description)
            .setPositiveButton(R.string.quest_name_noName_confirmation_positive) { _, _ -> applyAnswer(NoRoadName) }
            .setNegativeButton(R.string.quest_generic_confirmation_no, null)
            .show()
    }
}
